subsys/bootloader/fprotect/fprotect_bprot.c

```diff
@@ -4,7 +4,62 @@
  * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
  */
 
-void fprotect_area(u32_t start, size_t num_pages)
+#include <stddef.h>
+#include <nrf.h>
+#include <nrf_peripherals.h>
+#include <misc/util.h>
+#include <misc/__assert.h>
+#include <errno.h>
+
+#if !defined(NRF_BPROT) && defined(NRF_MPU)
+/* Rename nRF51's NRF_MPU to NRF_BPROT as they work the same. */
+#define NRF_BPROT NRF_MPU
+#define CONFIG0 PROTENSET0
+#define CONFIG1 PROTENSET1
+#endif
+
+/* The number of CONFIG registers present in the chip. */
+#define BPROT_CONFIGS_NUM ceiling_fraction(BPROT_REGIONS_NUM, BITS_PER_LONG)
+
+int fprotect_area(u32_t start, size_t length)
 {
-	/*  TODO - no BPROT HAL available. */
+	/*  TODO - use BPROT HAL when available. */
+
+	u32_t pagenum_start = start / BPROT_REGIONS_SIZE;
+	u32_t pagenum_end   = (start + length) / BPROT_REGIONS_SIZE;
+
+	if ((start % BPROT_REGIONS_SIZE) ||
+	    (length % BPROT_REGIONS_SIZE) ||
+	    (pagenum_end >= BPROT_REGIONS_NUM)) {
+		/* start or length isn't aligned with a BPROT region,
+		 * or attempting to protect an area outside flash. */
+		return -EFAULT;
+	}
+
```

Comment:
`if (pagenum_end <= pagenum_start) return -EINVAL`?

Reply:
Good catch :). Changed retval to -EINVAL and added the check.